lib/io/hardware_gpio_alt.cpp

```diff
@@ -0,0 +1,223 @@
+#include <unistd.h>
+
+namespace hyped::io {
+
+HardwareGpioReader::HardwareGpioReader(core::ILogger &log, const int read_file_descriptor)
+    : logger_(log),
+      read_file_descriptor_(read_file_descriptor)
+{
+}
+
+HardwareGpioReader::~HardwareGpioReader()
+{
+  close(read_file_descriptor_);
+}
+
+std::optional<core::DigitalSignal> HardwareGpioReader::readPin()
+{
+  // Read the value from the file
+  char read_buffer[2];
+  const auto offset = lseek(read_file_descriptor_, 0, SEEK_SET);
+  if (offset != 0) {
+    logger_.log(core::LogLevel::kFatal, "Failed to reset file offset");
+    return std::nullopt;
+  }
+  const auto read_result = read(read_file_descriptor_, read_buffer, sizeof(read_buffer));
+  if (read_result != sizeof(read_buffer)) {
+    logger_.log(core::LogLevel::kFatal, "Failed to read GPIO value");
+    return std::nullopt;
+  }
+  // Convert the read value to a DigitalSignal
+  const int value = std::atoi(read_buffer);
+  if (value == 0) {
+    return core::DigitalSignal::kLow;
+  } else if (value == 1) {
+    return core::DigitalSignal::kHigh;
+  } else {
+    logger_.log(core::LogLevel::kFatal, "Invalid GPIO value read");
+    return std::nullopt;
+  }
+}
+
+HardwareGpioWriter::HardwareGpioWriter(core::ILogger &log, const int write_file_descriptor)
+    : logger_(log),
+      write_file_descriptor_(write_file_descriptor)
+{
+}
+
+HardwareGpioWriter::~HardwareGpioWriter()
+{
+  close(write_file_descriptor_);
+}
+
+core::Result HardwareGpioWriter::writeToPin(const core::DigitalSignal state)
+{
+  // Convert DigitalSignal to a string
+  const std::uint8_t signal_value = static_cast<std::uint8_t>(state);
+  char write_buffer[2];
+  snprintf(write_buffer, sizeof(write_buffer), "%d", signal_value);
+  // Write the value to the file
+  const auto write_result = write(write_file_descriptor_, write_buffer, sizeof(write_buffer));
+  if (write_result != sizeof(write_buffer)) {
+    logger_.log(core::LogLevel::kFatal, "Failed to write GPIO value");
+    return core::Result::kFailure;
+  }
+  logger_.log(core::LogLevel::kDebug, "Wrote %d to GPIO", signal_value);
+  return core::Result::kSuccess;
+}
+
+HardwareGpio::HardwareGpio(core::ILogger &log) : logger_(log)
+{
+}
+
+std::optional<std::shared_ptr<IGpioReader>> HardwareGpio::getReader(const std::uint8_t pin,
+                                                                    const Edge edge)
+{
+  const auto initialise_result = initialisePin(pin, edge, Direction::kIn);
+  if (initialise_result == core::Result::kFailure) {
+    logger_.log(core::LogLevel::kFatal, "Failed to initialise GPIO %d", pin);
+    return std::nullopt;
+  }
+  const int read_file_descriptor = getFileDescriptor(pin, Direction::kIn);
+  if (read_file_descriptor < 0) {
+    logger_.log(core::LogLevel::kFatal, "Failed to get file descriptor for GPIO %d", pin);
+    return std::nullopt;
+  }
+  HardwareGpioReader reader(logger_, read_file_descriptor);
+  return std::make_shared<HardwareGpioReader>(reader);
```

Comment:
i want to comment on something here, for your understanding; feel free to resolve after reading (RAR)

The "right" way to do this would be `return std::make_shared<...>(logger_, read_file_descriptor);`. Of course, this doesn't work because the constructor is private so that wouldn't work. 

How could we resolve this? Option 1 is to use nested classes and make the reader/write a private class inside the Gpio. Then we could make the constructor public (and within the `HardwareGpio` class). We like to avoid nested classes so we accept the ugliness of this. Option 2 would be to make the constructor public. This usually isn't a problem if we assume that the user understands the design. However, this is a bold assumption.

Besides style, why would the single line way be better? Well, we avoid an unnecessary copy. Then again, these sorts of functions are only called on initialisation (i.e. O(1) times) so we don't really care about performance here.

Reply:
i see, do appreciate this btw. not always sure what the "right" way to do things even are